# Проект по SQL - Анализ книжной базы данных

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.<br>
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.<br>
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.<br>

### Описание данных
Таблица `books`<br>
Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

Таблица `authors`<br>
Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

Таблица `publishers`<br>
Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

Таблица `ratings`<br>
Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

Таблица `reviews`<br>
Содержит данные о пользовательских обзорах:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя автора обзора;
- `text` — текст обзора.

<image src="https://i.pstorage.space/i/61axvbl6M/original_books_sql.png" alt="">


### Ход исследования
Исследование пройдёт в три этапа:
1. Загрузка данных и изучение общей информации
2. Исследовательский анализ данных
3. Общий вывод

## 1. Загрузка данных и изучение общей информации

In [2]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**Таблица `books`**

In [3]:
query = '''SELECT * FROM books'''
con=engine.connect()
books = pd.io.sql.read_sql(sql=text(query), con = con)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


**Таблица `authors`**

In [5]:
query = '''SELECT * FROM authors'''
con=engine.connect()
authors = pd.io.sql.read_sql(sql=text(query), con = con)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


**Таблица `publishers`**

In [7]:
query = '''SELECT * FROM publishers'''
con=engine.connect()
publishers = pd.io.sql.read_sql(sql=text(query), con = con)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


**Таблица `ratings`**

In [9]:
query = '''SELECT * FROM ratings'''
con=engine.connect()
ratings = pd.io.sql.read_sql(sql=text(query), con = con)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


**Таблица `reviews`**

In [11]:
query = '''SELECT * FROM reviews'''
con=engine.connect()
reviews = pd.io.sql.read_sql(sql=text(query), con = con)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [12]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


## 2. Исследовательский анализ данных

**1) Посчитаем сколько книг вышло после 1 января 2000 года.**

In [13]:
query = '''SELECT COUNT(book_id) AS book_count
           FROM books
           WHERE CAST (publication_date AS timestamp) > '2000-01-01'
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_count
0         819

- После 1 января 2000 года вышло 819 книг.

**2) Для каждой книги посчитаем количество обзоров и среднюю оценку.**

In [21]:
query = '''SELECT b.book_id,
                  b.title AS book_title,
                  COUNT(DISTINCT rv.review_id) AS reviews_count,
                  AVG(rating) AS rating_avg
           FROM books AS b
           LEFT JOIN reviews AS rv ON rv.book_id = b.book_id
           LEFT JOIN ratings AS rt ON rt.book_id = rv.book_id
           GROUP BY b.book_id
           ORDER BY b.book_id
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                         book_title  \
0          1                                       'Salem's Lot   
1          2                 1 000 Places to See Before You Die   
2          3  13 Little Blue Envelopes (Little Blue Envelope...   
3          4  1491: New Revelations of the Americas Before C...   
4          5                                               1776   
..       ...                                                ...   
995      996           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997                        Xenocide (Ender's Saga  #3)   
997      998                                    Year of Wonders   
998      999                        You Suck (A Love Story  #2)   
999     1000  Zen and the Art of Motorcycle Maintenance: An ...   

     reviews_count  rating_avg  
0                2    3.666667  
1                1    2.500000  
2                3    4.666667  
3                2    4.500000  
4                4    4.000000  
..             ...         ...  
995              3    3.666667  
996              3    3.400000  
997              4    3.200000  
998              2    4.500000  
999              4    3.833333  

[1000 rows x 4 columns]

- У нас в датабазе тысяча книг и перед нами весь их список с артикулом, названием, количеством пользовательских обзоров и средним пользовательским рейтингом.
- Самое большое количество обзоров сделано на "Twilight" в количестве 7 штук.
- Самый высокий рейтинг - 5, его получили 43 книги.
- Самый низкий рейтинг - 1,5, его получила книга "Harvesting the Heart".

**3) Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры.**

In [15]:
query = '''SELECT p.publisher_id,
                  p.publisher,
                  COUNT(b.book_id) AS book_count
           FROM books AS b
           LEFT JOIN publishers AS p ON p.publisher_id = b.publisher_id
           WHERE b.num_pages > 50
           GROUP BY p.publisher_id
           ORDER BY book_count DESC
           LIMIT 1
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id      publisher  book_count
0           212  Penguin Books          42

- Наибольшее число книг толще 50 страниц выпустило издательство "Penguin Books" - 42 наименования.

**4) Определим автора с самой высокой средней оценкой книг — будем учитывать только книги с 50 и более оценками.**

In [16]:
query = '''SELECT a.author_id, 
                  a.author,
                  AVG(r.rating) AS rating_avg
           FROM books AS b
           LEFT JOIN authors AS a ON b.author_id = a.author_id
           LEFT JOIN ratings AS r ON b.book_id = r.book_id
           WHERE b.book_id IN (SELECT book_id
                               FROM ratings
                               GROUP BY book_id
                               HAVING COUNT(*) >= 50)
           GROUP BY a.author_id
           ORDER BY rating_avg DESC
           LIMIT 1
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                      author  rating_avg
0        236  J.K. Rowling/Mary GrandPré    4.287097

- Автор с самой высокой средней оценкой книг (среди книг с 50 и более оценками) - Джоан Роулинг в совторстве с иллюстратором Мэри Гранпрэ, средний рейтинг 4,29.

**5) Посчитаем среднее количество обзоров от пользователей, которые поставили больше 48 оценок.**

In [17]:
query = '''SELECT AVG(reviews_count)
           FROM (SELECT username,
                        COUNT(*) AS reviews_count
                 FROM reviews
                 WHERE username IN (SELECT username
                                    FROM ratings
                                    GROUP BY username
                                    HAVING COUNT(*) > 48)
                 GROUP BY username) AS usernames_reviews
           '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

-  Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.0

## 3. Общий вывод

Мы подключились к базе данных и из пяти представленных таблиц получили следующую информацию:
- После 1 января 2000 года вышло 819 книг.
- В датабазе 1000 книг.
- Самое большое количество обзоров сделано на "Twilight" в количестве 7 штук.
- Самый высокий рейтинг - 5, его получили 43 книги.
- Самый низкий рейтинг - 1,5, его получила книга "Harvesting the Heart".
- Наибольшее число книг толще 50 страниц выпустило издательство "Penguin Books" - 42 наименования.
- Автор с самой высокой средней оценкой книг (среди книг с 50 и более оценками) - Джоан Роулинг в совторстве с иллюстратором Мэри Гранпрэ, средний рейтинг 4,29.
- Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.0